MetadataExtractor/Formats/QuickTime/QuickTimeMetadataReader.cs

```diff
@@ -113,6 +115,69 @@ void UuidHandler(AtomCallbackArgs a)
                 }
             }
 
+            void MetaDataHandler(AtomCallbackArgs a)
+            {
+                // see https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html
+                switch (a.TypeString)
+                {
+                    case "keys":
+                    {
+                        var version = a.Reader.GetByte();
+                        var flags = a.Reader.GetBytes(3);
+                        var entryCount = a.Reader.GetUInt32();
+                        for (int i = 1; i <= entryCount; i++)
+                        {
+                            var keySize = a.Reader.GetUInt32();
+                            var keyValueSize = (int)keySize - 8;
+                            var keyNamespace = a.Reader.GetUInt32();
+                            var keyValue = a.Reader.GetBytes(keyValueSize);
+                            metaDataKeys.Add(Encoding.UTF8.GetString(keyValue));
+                        }
+                        break;
+                    }
+                    case "ilst":
+                    {
+                        var directory = new QuickTimeMetadataHeaderDirectory();
+                        // Iterate over the list of Metadata Item Atoms.
+                        for (int i = 0; i < metaDataKeys.Count; i++)
+                        {
+                            long atomSize = a.Reader.GetUInt32();
+                            if (atomSize < 24)
+                            {
+                                directory.AddError("Invalid ilist AtomSize");
+                                break;
+                            }
+                            var atomType = a.Reader.GetUInt32();
+
+                            // Indexes into the metadata item keys atom are 1-based (1�entry_count).
+                            // atom type for each metadata item atom is the index of the key
+                            if (atomType < 1 || atomType > metaDataKeys.Count)
+                            {
+                                directory.AddError("Invalid ilist AtomType");
+                                break;
+                            }
+                            var key = metaDataKeys[(int)atomType - 1];
+
+                            // Value Atom
+                            var typeIndicator = a.Reader.GetBytes(4);
+                            var localeIndicator = a.Reader.GetBytes(4);
+
+                            // Data Atom
+                            var dataTypeIndicator = a.Reader.GetBytes(4);
+                            var dataLocaleIndicator = a.Reader.GetBytes(4);
```

Comment:
Looking at the spec, it seems that these values cannot be ignored. The value is not guaranteed to be a UTF-8 string. The type indicator is packed to contain a [well-known type](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html#//apple_ref/doc/uid/TP40000939-CH1-SW28) field which instructs the encoding to use for the data value (e.g. UTF-8, UTF-16, etc...).

Consider reading `int32` values rather than `byte[]` as the former will not cause heap allocations.

We don't have any built-in support for per-value locale support. It may be interesting to add, though this would be the only usage I know of currently.

Reply:
I took some of my assumptions from [Table 3-6](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html#//apple_ref/doc/uid/TP40000939-CH1-SW43) (though this would leave `Artwork `and `UserRating` improperly handled).
Are you looking for a full implementation of `Well-Known Types` or would you be okay with providing support for types 1, 13, 14, 23 and 27 and treating other values as errors?

As for locale, I was hoping to provide an incremental improvement by adding support for GPS location retrieval, All of the samples that I had access to had this value set to `\0\0` so I did not have a need for it at the time.

Reply:
> would you be okay with providing support for types 1, 13, 14, 23 and 27 and treating other values as errors?

Yes that's a long table and I don't think we need to handle everything. Adding the framework to support different encodings and covering the ones we think would be most common and easiest is great for now. Having even basic support now lowers the bar for others who stumble upon data that's not covered.

I would avoid treating unsupported encodings as errors. Maybe just store the raw bytes instead. That's more in line with what's done elsewhere.